### PR TITLE
Add a test which calls a user-defined function with NULL argument

### PR DESCRIPTION
**Pull request by willmurnane**
I found this incorrect behavior, but I haven't figured out how to fix it.

### CODE REVIEW

File: pgx-utils/src/lib.rs [hunk 1]
```diff
@@ -129,8 +129,10 @@ pub enum FunctionArgs {
 
 #[derive(Debug)]
 pub enum CategorizedType {
-    Iterator(Vec<String>),
-    OptionalIterator(Vec<String>),
+    Iterator {
+        types: Vec<String>,
+        is_optional: bool,
+    },
```
Comment (workingjubilee) [inline pgx-utils/src/lib.rs:135]: I'm not sure I understand why we're now adding a bool into the same variant, as opposed to using a different variant, as previously done?

File: pgx-utils/src/rewriter.rs [hunk 2]
```diff
@@ -718,21 +694,18 @@ impl FunctionSignatureRewriter {
...
                             quote_spanned! {ident.span()=>
-                                let #name = pgx::pg_getarg::<#option_type>(#fcinfo_ident, #i);
+                                let #name : #type_ = pgx::pg_getarg::<#option_type>(#fcinfo_ident, #i);
                             }
                         } else if type_matches(&type_, "pg_sys :: FunctionCallInfo")
                             || type_matches(&type_, "pgx :: pg_sys :: FunctionCallInfo")
                         {
                             quote_spanned! {ident.span()=>
-                                let #name = #fcinfo_ident;
+                                let #name: pgx::pg_sys::FunctionCallInfo  = #fcinfo_ident;
```
Comment (workingjubilee) [inline pgx-utils/src/rewriter.rs:708]: rustfmt often leaves formatting in macro contexts untouched. Here, `#name:` is used, elsewhere, `#name :` is used, this must be consistent, and preferably consistent with the default type annotation formatting that rustfmt emits.

File: pgx-utils/src/rewriter.rs [hunk 3]
```diff
@@ -566,8 +537,18 @@ impl PgGuardRewriter {
                         if suffix_arg_name && ident.ident.to_string() != "fcinfo" {
                             let ident = Ident::new(&format!("{}_", ident.ident), ident.span());
                             arg_list.extend(quote! { #ident, });
+                            if type_matches(ty.ty.as_ref(), "()")
+                                || extract_type_from_option(ty.ty.as_ref()).is_some()
+                                || type_matches(&ty.ty, "pg_sys :: FunctionCallInfo")
+                                || type_matches(&ty.ty, "pgx :: pg_sys :: FunctionCallInfo")
```
Comment (workingjubilee) [inline pgx-utils/src/rewriter.rs:543]: Why the spaces in `pgx :: pg_sys :: FunctionCallInfo`?

File: pgx-utils/src/rewriter.rs [hunk 4]
```diff
@@ -482,7 +449,7 @@ impl PgGuardRewriter {
             func.sig.ident.span(),
         );
 
-        let arg_list = PgGuardRewriter::build_arg_list(&sig, false);
+        let (arg_list, _matchable) = PgGuardRewriter::build_arg_list(&sig, false);
```
Comment (workingjubilee) [inline pgx-utils/src/rewriter.rs:452]: This function had its return signature altered but the latter half of the tuple is used in only one location? Is there another way to do this?

**Review by workingjubilee — COMMENTED** [review]
Apologies for the delay in review. I find the content of this segment of the code hard to follow in general as I am not as experienced in it, and as I start working on attempting to update and land #615, it's likely to induce conflicts with this set of changes. So my initial remarks are just from a relatively cursory pass... in general this has proven to be an extremely touchy segment of the code, in the past, especially around SRFs, so any changes touching it could stand to add some comments and focus, a lot, on legibility.

File: pgx-utils/src/lib.rs (hunk 1 above)
Comment (willmurnane) [reply pgx-utils/src/lib.rs:135]: The different variant was translated into "call the same code, but with a boolean flag" anyways. I don't mind reverting this change, but it collapsed the code in `item_fn_with_rewrite` quite a bit. Combined with factoring out the `if types.len() == 1` check, that collapsed all 4 cases down to one code path.

File: pgx-utils/src/rewriter.rs (hunk 3 above)
Comment (willmurnane) [reply pgx-utils/src/rewriter.rs:543]: I think that's due to how [`format!("{}", quote!`](https://github.com/tcdi/pgx/blob/f2c54dc38e6b28cb7dcf43ef9f84df36e7108278/pgx-utils/src/rewriter.rs#L750) works. I was just matching the style seen elsewhere.

File: pgx-utils/src/rewriter.rs (hunk 4 above)
Comment (willmurnane) [reply pgx-utils/src/rewriter.rs:452]: I'm not sure it's entirely correct that I'm ignoring it in this case. I'm not totally clear on how `PgGuardRewriter .item_fn` is called and for what reasons. It looks like it's usually called from [here](https://github.com/tcdi/pgx/blob/develop/pgx-macros/src/lib.rs#L592-L598), where `rewrite_args` is always supplied as `false`, but is also called from [here](https://github.com/tcdi/pgx/blob/develop/pgx-macros/src/lib.rs#L46), where it's also hard-coded to a value of `false`. So based on I think it might be okay to entirely remove `item_fn_with_rewrite`, the argument `rewrite_args`, and so on. I didn't want to do that surgery without understanding more, though.

File: pgx-utils/src/rewriter.rs (hunk 2 above)
Comment (willmurnane) [reply pgx-utils/src/rewriter.rs:708]: Good catch. I'll update.

**Comment by willmurnane** [comment]
I agree with the overall challenging nature of this code, especially given the behavior of Rust macros ("[macro expansion to text is a lossy process](https://github.com/dtolnay/cargo-expand#disclaimer)"). I could try a couple of things to make this easier to review:
* break this PR into smaller, more digestible chunks that do exactly one thing
* add more comments: what is happening when this code is called, and why does (this) need to happen in response
* rebase onto #615
* dive into some of the major surgery I mentioned in review comments above.
I would be happy to do any or all of these. Let me know what would be helpful.

File: pgx-utils/src/rewriter.rs (hunk 4 above)
Comment (willmurnane) [reply pgx-utils/src/rewriter.rs:452]: Oops - I misread. In the first case, it's got a value of `true` hard-coded, not `false`. Not sure how I managed that one...

File: pgx-utils/src/lib.rs (hunk 1 above)
Comment (workingjubilee) [reply pgx-utils/src/lib.rs:135]: I would actually expect it to be such an... "overly ritualistic" form? It does suggest this code needs to be refactored, but there's something in the back of my mind that suggests this particular refactoring will reach a local maxima faster and thus lower than other ones which try to address the code factoring from a "one-level up" perspective.

**Comment by workingjubilee** [comment]
I think this is fine as one PR, but comments comments comments.

I don't think you want to rebase onto it yet. Soon maybe. If you want to give it a quick experimental pass on your local to see if there are any conflicts you think are particularly noteworthy, sure, but that PR itself needs to be updated.

**Review by workingjubilee — CHANGES_REQUESTED** [review]
I want to land this before 0.5.0, truly I do, but this PR too is also dashed upon the rocks of #615.

**Comment by workingjubilee** [comment]
#615 has landed.

**Comment by willmurnane** [comment]
#615 fixed the problem that this PR was addressing. I've winnowed it down to just add the test that previously failed.

**Review by workingjubilee — APPROVED** [review]
Which also makes it much easier to review. Nice!
